server.go

```diff
@@ -187,3 +224,196 @@ func copyWithHeader(r io.Reader, w io.Writer, header proto.Header) error {
+	}
+
+	err = os.WriteFile(config, []byte(strings.Join(settings, "\n")), 0o644)
+	if err != nil {
+		return nil, xerrors.Errorf("unable to create %s for session %s: %w", config, id, err)
+	}
+
+	var sess *session
+	if rawSession, ok := sessions.Load(id); ok {
+		if sess, ok = rawSession.(*session); !ok {
+			return nil, xerrors.Errorf("found invalid type in session map for ID %s", id)
+		}
+	} else {
+		ctx, cancel := context.WithCancel(context.Background())
+		sess = &session{
+			ctx:   ctx,
+			ready: make(chan error, 1),
+		}
+		sessions.Store(id, sess)
+
+		// Starting screen with -Dm causes it to launch a server tied to this
+		// process, letting us attach to and kill it with the PID.
+		process, err := execer.Start(ctx, Command{
+			Command:    "screen",
+			Args:       append([]string{"-S", id, "-Dmc", config, command.Command}, command.Args...),
+			UID:        command.UID,
+			GID:        command.GID,
+			Env:        append(command.Env, "SCREENDIR="+screendir),
+			WorkingDir: command.WorkingDir,
+		})
+		if err != nil {
+			cancel()
+			err = xerrors.Errorf("failed to create session %s: %w", id, err)
+			sess.ready <- err
+			close(sess.ready)
+			return nil, err
+		}
+
+		sess.screenID = fmt.Sprintf("%d.%s", process.Pid(), id)
+
+		// Timeouts created with AfterFunc can be reset.
+		sess.timeout = time.AfterFunc(options.SessionTimeout, func() {
+			// Delete immediately in case it takes a while to clean up otherwise new
+			// connections with this ID will try to connect to the closing screen.
+			sessions.Delete(id)
+			// Close() will send a SIGTERM allowing screen to clean up.  If we kill it
+			// abruptly the socket will be left behind which is probably harmless but
+			// it causes screen -list to show a bunch of dead sessions (of course the
+			// user would not see this since we have a custom socket directory) and it
+			// seems ideal to let it clean up anyway.
+			process.Close()
+			select {
+			case <-sess.ctx.Done():
+				return
+			case <-time.After(5 * time.Second):
+				cancel() // Force screen to exit.
+			}
+		})
+
+		// Remove the session if screen exits.
+		go func() {
+			err := process.Wait()
+			// Screen exits with a one when it gets a SIGTERM.
+			if exitErr, ok := err.(ExitError); ok && exitErr.ExitCode() != 1 {
+				flog.Error("session %s exited with error %w", id, err)
+			}
+			cancel()
+			sess.timeout.Stop()
+			sessions.Delete(id)
+		}()
+
+		// Sometimes if you attach too quickly after spawning screen it will say the
+		// session does not exist so run a command against it until it works.
+		go func() {
+			ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+			defer cancel()
+			defer close(sess.ready)
+			for {
+				select {
+				case <-sess.ctx.Done():
+					sess.ready <- xerrors.Errorf("session %s is gone", id)
+					return
+				case <-ctx.Done():
+					sess.ready <- xerrors.Errorf("timed out waiting for session %s", id)
+					return
```

Comment:
I've seen this pattern in multiple places where we return a single error on the channel then close it.  That means the first reader will get the error, but any subsequent readers will get `nil` which they interpret as no error.  This will give us trouble with multiple reconnection attempts.  I'd prefer a more robust solution that registers the error on the session and returns it to all readers.

Reply:
Ah crap I thought it always returned the last value but it seems I wildly misread the docstring.  Will refactor.